src/engraving/libmscore/segment.cpp

```diff
@@ -2706,28 +2706,63 @@ bool Segment::hasAccidentals() const
     return false;
 }
 
-CrossStaffContent Segment::crossStaffContent() const
-{
-    CrossStaffContent crossStaffContent;
-    for (EngravingItem* el : elist()) {
-        if (el && el->isChordRest()) {
-            if (toChordRest(el)->staffMove() == 0 && !toChordRest(el)->isFullMeasureRest()) {
-                crossStaffContent.movedDown = false;
-                crossStaffContent.movedUp = false;
-                break;
+/************************************************************************
+ * computeCrossBeamType
+ * Looks at the chords of this segment and next segment to detect beams
+ * with alternating stem directions (upDown: this chord is up, next chord
+ * is down; downUp: this chord is down, next chord is up). Needed for
+ * correct horizontal spacing of cross-beam situations.
+ * **********************************************************************/
+
+void Segment::computeCrossBeamType(Segment* nextSeg)
+{
+    _crossBeamType.reset();
+    if (!isChordRestType() || !nextSeg || !nextSeg->isChordRestType()) {
+        return;
+    }
+    bool upDown = false;
+    bool downUp = false;
+    for (EngravingItem* e : elist()) {
+        if (!e || !e->isChordRest() || !e->staff()->visible()) {
+            continue;
+        }
+        ChordRest* thisCR = toChordRest(e);
+        if (!thisCR->visible() || thisCR->isFullMeasureRest()) {
+            continue;
+        }
+        if (!thisCR->beam()) {
+            return;
```

Comment:
Just to be sure, is it intended that in this case, _crossBeamType stays `{false, false}` even though `upDown` or `downUp` might be true? (I guess this will mean that if on at least _one_ staff there is no "cross beam", no adjustment will be made, is that right?)

Also, it looks like in the following case, the `_crossBeamType` of the first segment is `{false, false}`, so the spacing between the first two eight notes won't be adjusted. I suppose that this is intentional for now, and we might refine the algorithm even further in the future?
<img width="374" alt="Scherm­afbeelding 2022-09-02 om 17 40 21" src="https://user-images.githubusercontent.com/48658420/188189066-f5d1d976-3775-4b29-91e7-5f3cb84ff556.png">

Reply:
Quick answer: yes. Adjusting for cross-beams should never interfere with spacing of "normal" notes in any other voice or staff, so I'm being very strict and I'm not applying any correction if there is even a single non-cross-beam note anywhere in the same segment or the next. Second measure is as expected here, first measure could be improved. I think I'll refine it at a later stage. For now, I'd say it can go as it is.